src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1628,42 +1628,61 @@
     (fs/delete csr-path)
     (log/debug (i18n/trs "Removed certificate request for {0} at ''{1}''" subject csr-path))))
 
+(schema/defn filter-already-revoked-serials :- [schema/Int]
+  "Given a list of serials and Puppet's CA CRL, returns vector of serials with
+   any already-revoked serials removed."
+  [serials :- [schema/Int]
+   crl :- X509CRL]
+  (let [crl-revoked-list (.getRevokedCertificates crl)
+        existed-serials (set (map #(.getSerialNumber %) crl-revoked-list))
+        duplicate-serials (set/intersection (set serials) existed-serials)] ; For info output
+        (when (> (count duplicate-serials) 0)
+          (doseq [serial duplicate-serials]
+            (log/debug (str "Certificate with serial " serial " is already revoked.")))) ; Feedback on where to place the info output.
+        (vec (set/difference (set serials) existed-serials))))
+
 (schema/defn revoke-existing-certs!
   "Revoke the subjects' certificates. Note this does not destroy the certificates.
    The certificates will remain in the signed directory despite being revoked."
   [{:keys [signeddir cacert cacrl cakey infra-crl-path
            infra-node-serials-path enable-infra-crl]} :- CaSettings
    subjects :- [schema/Str]]
-  (let [serials (map #(-> (path-to-cert signeddir %)
-                       (utils/pem->cert)
-                       (utils/get-serial))
-                  subjects)
-        serial-count (count serials)
-        [our-full-crl & rest-of-full-chain] (utils/pem->crls cacrl)
-        new-full-crl (utils/revoke-multiple our-full-crl
+  (let [[our-full-crl & rest-of-full-chain] (utils/pem->crls cacrl)
+        serials (filter-already-revoked-serials (map #(-> (path-to-cert signeddir %)
+                                                          (utils/pem->cert)
+                                                          (utils/get-serial))
+                                                          subjects) our-full-crl)
+        serial-count (count serials)]
+    (if (= 0 serial-count)
+      (log/info "No revoke action needed.  The certs are already in revoked list.") ; Info output here?
```

Comment:
I guess we should probably also wrap all these log messages with `i18n/trs`, since we do that in the existing log message.
(This is for potential translation. `trsn` is used when there are different forms of the string depending on some count. Most of these messages don't need that, so `trs` is sufficient.)